tianshou/policy/base.py

```diff
@@ -511,7 +530,8 @@ def update(
         batch, indices = buffer.sample(sample_size)
         self.updating = True
         batch = self.process_fn(batch, buffer, indices)
-        training_stat = self.learn(batch, **kwargs)
+        with in_train_mode(self):
```

Comment:
shouldn't update or rather learn temporary set the `is_within_training_step` to True?

@opcode81 @maxhuettenrauch

Reply:
Ah, now I see, it's the correct behavior. We should still adjust it to ensure that it's set to False -with a private contextmanager?

In the current implementation in PGPolicy and so, it has the logic of `within_training_rollout`, where we want to sample from the dist. Here we want to compute gradients, so we are no longer within a rollout.

We are still within a training iteration, so I feel like the name is confusing. How about `within_training_rollout` or `within_training_data_collection`?

Reply:
No, `is_within_training_step` has precisely the semantics we want:

![image](https://github.com/thu-ml/tianshou/assets/1295099/81ecb7aa-6b51-49f3-93f8-3f35420eb0bf)

The term **training step** should become part of a _glossary_ of well-defined terms in Tianshou.
It's already used in, for example, SamplingConfig - and now it is also reflected in the Trainer (method `training_step`).

> We should still adjust it to ensure that it's set to False -with a private contextmanager?

This is very unclear. Adjust what, where, and why?

Reply:
After discussion: the resolution is to raise an error if the flag is false when update is called

Reply:
Done